Kernel/Devices/FUSEDevice.cpp

```diff
@@ -0,0 +1,201 @@
+        Optional<size_t> instance_index = {};
+        for (size_t i = 0; i < instances.size(); ++i) {
+            if (instances[i].fd == &fd) {
+                instance_index = i;
+                break;
+            }
+        }
+
+        if (!instance_index.has_value())
+            return Error::from_errno(ENODEV);
+
+        auto& instance = instances[instance_index.value()];
+
+        if (instance.expecting_header) {
+            memset(instance.response->data(), 0, instance.response->size());
+
+            fuse_out_header header;
+            TRY(buffer.read(&header, 0, sizeof(fuse_out_header)));
+
+            dbgln_if(FUSE_DEBUG, "header: length: {}, error: {}, unique: {}", header.len, header.error, header.unique);
+            memcpy(instance.response->data(), &header, sizeof(fuse_out_header));
+
+            if (header.len > sizeof(fuse_out_header))
+                instance.expecting_header = false;
+            else
+                instance.response_ready = true;
+        } else {
+            fuse_out_header* existing_header = bit_cast<fuse_out_header*>(instance.response->data());
+
+            instance.expecting_header = true;
+            if (existing_header->len > instance.response->size())
+                return Error::from_errno(EINVAL);
+
+            instance.response_ready = true;
+            u64 length = existing_header->len - sizeof(fuse_out_header);
+            dbgln_if(FUSE_DEBUG, "request: response length: {}", length);
+            TRY(buffer.read(instance.response->data() + sizeof(fuse_out_header), 0, length));
+        }
+
+        return size;
+    });
+}
+
+ErrorOr<NonnullOwnPtr<KBuffer>> FUSEDevice::send_request_and_wait_for_a_reply(OpenFileDescription const& description, Bytes bytes)
+{
+    MutexLocker locker(m_lock);
+    return m_instances.with([&](auto& instances) -> ErrorOr<NonnullOwnPtr<KBuffer>> {
+        Optional<size_t> instance_index = {};
+        for (size_t i = 0; i < instances.size(); ++i) {
+            if (instances[i].fd == &description) {
+                instance_index = i;
+                break;
+            }
+        }
+
+        VERIFY(instance_index.has_value());
+        auto& instance = instances[instance_index.value()];
+
+        VERIFY(!instance.drop_request);
+        VERIFY(bytes.size() <= 0x21000);
+
+        memset(instance.pending_request->data(), 0, instance.pending_request->size());
+        memcpy(instance.pending_request->data(), bytes.data(), bytes.size());
+        instance.buffer_ready = true;
+        evaluate_block_conditions();
+
+        while (!instance.response_ready)
+            (void)Thread::current()->sleep(Duration::from_microseconds(100));
+
+        auto result = KBuffer::try_create_with_bytes("FUSEDevice: Response"sv, instance.response->bytes());
+        instance.response_ready = false;
+
+        return result;
+    });
+}
+
+void FUSEDevice::shutdown_for_description(OpenFileDescription const& description)
+{
+    MutexLocker locker(m_lock);
```

Comment:
Do we still need the `m_lock` locking here? isn't locking `m_instances` enough to synchronize things?